### PR TITLE
link-grammar 5.10.2

### PR DESCRIPTION
**Pull request by chenrui333**
Created with `brew bump-formula-pr`.

### CODE REVIEW

**Comment by chenrui333** [comment]
```
Making install in python
  CXX      _clinkgrammar_la-lg_python_wrap.lo
  CXXLD    _clinkgrammar.la
 ../.././install-sh -c -d '/opt/homebrew/opt/python@3.9/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/linkgrammar'
 ../.././install-sh -c -d '/opt/homebrew/opt/python@3.9/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages'
 ../.././install-sh -c -d '/opt/homebrew/opt/python@3.9/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/linkgrammar'
 /usr/bin/install -c -m 644 ./linkgrammar.pth '/opt/homebrew/opt/python@3.9/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages'
install: /opt/homebrew/opt/python@3.9/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/linkgrammar.pth: Operation not permitted
make[3]: *** [install-pkgpypathPYTHON] Error 71
make[3]: *** Waiting for unfinished jobs....
mkdir: /opt/homebrew/opt/python@3.9/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/linkgrammar: Operation not permitted
mkdir: /opt/homebrew/opt/python@3.9/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/linkgrammar: Operation not permitted
make[3]: *** [install-pkgpyexecLTLIBRARIES] Error 1
make[3]: *** [install-pkgpythonPYTHON] Error 1
```

**Comment by carlocab** [comment]
This is trying to install a Python module into `python@3.9`'s prefix, which of course it won't be able to. We need to tell it to install the Python module into its own prefix.

**Comment by linas** [comment]
Hmm. This is the same install path that python3 has used for the last 8+ years; it has not changed. It's the recommended default install path on "conventional" python3 installations (to the best of my understanding).  I guess you are saying its different somehow on Apple?

I believe that link-grammar has not altered this install path since the inception; we're not doing anything different here, than before. So I'm guessing that something else changed, in some other package, that is having a side-effect of causing this to fail.

**Comment by linas** [comment]
Re: spdx: I suppose it would be "LGPL-2.1-or-later" based on my understanding of what the SPDX people want.

**Comment by carlocab** [comment]
> I guess you are saying its different somehow on Apple?

This is probably more a quirk of the way Homebrew installs things -- each package has its own prefix which are then symlinked into Homebrew's prefix. Not sure why this is only coming up now, though.

Is there a flag we can pass to your `configure` script that would allow us to specify this location? I had a quick look at this and I didn't see an obvious way of doing it, but perhaps I missed something.

**Comment by linas** [comment]
Link Grammar does nothing at all to control installation. This is done automagically by automake/autoconf. In other words, Link Grammar should install exactly the same way as any other package that is using automake/autoconf to install python things.  HOWEVER....

As I read the link-grammar `bindings/python/Makefile.am` I see this suspicious code:
```
# Over-ride the install location, so as to remove the dash in the
# directory "link-grammar".  Turns out python cannot tolerate dashes.
# Thus, typically, it will install into
# /usr/local/lib/python2.7/dist-packages/linkgrammar (without the dash)
pkgpythondir=$(pythondir)/linkgrammar
pkgpyexecdir=$(pythondir)/linkgrammar
```

It is possible that, on your system, `pkgpythondir` and `pkgpyexecdir` point to reasonable locations, and I just clobbered them with the unreasonable `pythondir`.  So it seems the correct fix is to remove the dash from `pkgpythondir` without otherwise clobbering it.  Not sure how to do that.  I guess with `sed`. I will need your help testing the fix.

**Comment by linas** [comment]
The install paths appear to be governed by these variables:
https://www.gnu.org/software/automake/manual/html_node/Python.html

I just now changed the `Makefile.am` as follows
```
--- a/bindings/python/Makefile.am
+++ b/bindings/python/Makefile.am
@@ -19,7 +19,7 @@ built_py_sources = $(top_builddir)/bindings/python/clinkgrammar.py
 # Thus, typically, it will install into
 # /usr/local/lib/python2.7/dist-packages/linkgrammar (without the dash)
 pkgpythondir=$(pythondir)/linkgrammar
-pkgpyexecdir=$(pythondir)/linkgrammar
+pkgpyexecdir=$(pyexecdir)/linkgrammar
 
 # Files that get installed in $pkgpythondir
 pkgpython_PYTHON =                                 \
```

Three ways to test: (1) ask me to give you a private tarball (2) hand-edit your `Makefile.in` file, circa line 388 and replace it so it says `pkgpyexecdir = $(pyexecdir)/linkgrammar`  (3) get the git source, and autoconf/autmake it.

I can't tell what might be easiest for you.

**Comment by linas** [comment]
ping

**Comment by linas** [comment]
FYI, link-grammar is exploring arm-related build problems in opencog/link-grammar#1247 .. I'll try to update here. FWIW: the actual problem is a failing unit test for some brand-new code that no one uses (except me) and no one is likely to use, unless they are doing some bleeding-edge linguistics research (which is very unlikely, and thus should not impact Homebrew.)  In other words: I believe that LG works great on ARM, and its blocked only by the above failing unit test. (But perhaps I am misunderstanding the openjdk issue???)

I think that maybe the above python-install-path-problem *might* be fixed in commit a613961489bbeaf4333a53c6139ab6a9adb363a9 maybe, but this needs to be tested.  If it is fixed, I will issues a version 5.10.2 that includes this fix (and includes ARM fixes, if those arrive in a timely manner)

I need confirmation that commit a613961489bbeaf4333a53c6139ab6a9adb363a9 fixes the python-install-path problem. If it does not, then ... I'm a bit lost, and will need to add debug prints to the code, and some assistance/guidance from you.

**Comment by cho-m** [comment]
Tried applying as patch, but same issue. From configure output, the path for both are the same,
```
checking for python3 prefix... /usr/local/opt/python@3.9/Frameworks/Python.framework/Versions/3.9
checking for python3 exec_prefix... /usr/local/opt/python@3.9/Frameworks/Python.framework/Versions/3.9
checking for python3 platform... darwin
checking for python3 script directory... ${PYTHON_PREFIX}/lib/python3.9/site-packages
checking for python3 extension module directory... ${PYTHON_EXEC_PREFIX}/lib/python3.9/site-packages
checking for python3.9... (cached) /usr/local/opt/python@3.9/bin/python3
checking for a version of Python >= '2.1.0'... yes
checking for a version of Python >= '3.4'... yes
checking for the distutils Python package... yes
checking for Python include path... -I/usr/local/opt/python@3.9/Frameworks/Python.framework/Versions/3.9/include/python3.9
checking for Python library path... -undefined dynamic_lookup
checking for Python site-packages path... /usr/local/opt/python@3.9/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages
```

We probably want to see if we can install into link-grammar's directory.

In other formulae, we often manually set `prefix/Language::Python.site_packages("python3")` during `configure` or `cmake`.

For original Homebrew prefix, this will be `/usr/local/Cellar/link-grammar/5.10.1/lib/python3.9/site-packages`, which will link into `/usr/local/lib/python3.9/site-packages` and be available for Homebrew's python3.9.

File: Formula/link-grammar.rb [hunk 1]
```diff
@@ -30,15 +31,14 @@ class LinkGrammar < Formula
...
-                          "--with-regexlib=c"
-    system "make", "install"
+    inreplace "bindings/python/Makefile.am", "$(PYTHON_LDFLAGS) -module -no-undefined",
+                                             "$(PYTHON_LDFLAGS) -module"
+    system "autoreconf", "--verbose", "--install", "--force"
+    system "./configure", *std_configure_args, "--with-regexlib=c"
+
+    site_packages = prefix/Language::Python.site_packages("python3")
+    system "make", "install", "pythondir=#{site_packages}",
+                              "pyexecdir=#{site_packages}"
```
Comment (cho-m) [inline Formula/link-grammar.rb:41]: Only overriding values at install-time seemed to work (either `pythondir` like above, or `PYTHON_PREFIX=#{prefix}`) while environment variables & configure-time flags didn't help.

May be better to add a `configure` flag to override this value so all generated Makefiles are set to requested path. I don't know what is recommended way to do this, but some examples from other formulae include:
- graph-tool has configure flag `--with-python-module-path`
- bind has configure  flag `--with-python-install-dir`
- stp has cmake variable `PYTHON_LIB_INSTALL_DIR`

**Comment by carlocab** [comment]
@linas I wouldn't worry about where the Python bindings get installed -- the issues we're having here are peculiar to the way Homebrew installs things. It might be nice to expose `pythondir` and/or `pyexecdir` as `configure` options, though.

I am concerned about having to remove the `-no-undefined` flag in your `Makefile.am`, however.

**Comment by linas** [comment]
Link Grammar does nothing at all to control installation. This is done automagically by automake/autoconf. In other words, Link Grammar should install exactly the same way as any other package that is using automake/autoconf to install python things.  Are there any other packages in Homebrew that use autoconf to install python modules? What do they do?

The install paths appear to be governed by these variables:
https://www.gnu.org/software/automake/manual/html_node/Python.html
It does not mention how to over-ride those settings. I am looking to see how to over-ride `pythondir` and `pyexecdir` right now. 

The `-no-undefined` flag, as I understand it, forces a complete linking of the python module, so that, when it is loaded at runtime, it won't fail with unresolved symbols. Apparently, the python loader will not invoke the libc loader to do this.  It is possible that this is not an issue on Apple/Homebrew, so it might be safe ... 

Link-grammar does NOT have have any post-install unit tests, to see if it works after the install... it probably would be wise to have these for python, but I've never looked into this before.

**Comment by linas** [comment]
p.s. I will spin a version 5.10.2 shortly, which fixes some 32-bit and ARM issues.

**Comment by linas** [comment]
So I tried this:
```
make install pyexecdir=/foo
```
and I get the following error:
```
make[3]: Entering directory '/home/linas/src/novamente/src/link-grammar-git/build/bindings/python'
 /bin/mkdir -p '/foo/linkgrammar'
 /bin/bash ../../libtool   --mode=install /usr/bin/install -c   _clinkgrammar.la '/foo/linkgrammar'
libtool:   error: error: cannot install '_clinkgrammar.la' to a directory not ending in /usr/local/lib/python3.9/site-packages/linkgrammar
```

in other words, If I overload the install location, `libtool` will refuse to install to that location. ... Can you try installing like that?

**Comment by linas** [comment]
However, saying 
```
make install pythondir=/foo
```

does work, and everything got istalled into the directory `foo`:
```
$ find /foo
/foo
/foo/linkgrammar
/foo/linkgrammar/clinkgrammar.py
/foo/linkgrammar/linkgrammar.py
/foo/linkgrammar/__pycache__
/foo/linkgrammar/__pycache__/__init__.cpython-39.pyc
/foo/linkgrammar/__pycache__/__init__.cpython-39.opt-1.pyc
/foo/linkgrammar/__pycache__/linkgrammar.cpython-39.pyc
/foo/linkgrammar/__pycache__/clinkgrammar.cpython-39.pyc
/foo/linkgrammar/__pycache__/clinkgrammar.cpython-39.opt-1.pyc
/foo/linkgrammar/__pycache__/linkgrammar.cpython-39.opt-1.pyc
/foo/linkgrammar/__init__.py
/foo/linkgrammar.pth
$ find /foo
/foo
/foo/linkgrammar
/foo/linkgrammar/clinkgrammar.py
/foo/linkgrammar/linkgrammar.py
/foo/linkgrammar/__pycache__
/foo/linkgrammar/__pycache__/__init__.cpython-39.pyc
/foo/linkgrammar/__pycache__/__init__.cpython-39.opt-1.pyc
/foo/linkgrammar/__pycache__/linkgrammar.cpython-39.pyc
/foo/linkgrammar/__pycache__/clinkgrammar.cpython-39.pyc
/foo/linkgrammar/__pycache__/clinkgrammar.cpython-39.opt-1.pyc
/foo/linkgrammar/__pycache__/linkgrammar.cpython-39.opt-1.pyc
/foo/linkgrammar/__init__.py
/foo/linkgrammar.pth
```

**Comment by linas** [comment]
... and version 5.10.2 is now out ... it includes commit a613961489bbeaf4333a53c6139ab6a9adb363a9 which has a very slim but unlikely chance of maybe fixing this issue.

**Comment by linas** [comment]
Wow! the failures are because the ftp upload was borked!  I just fixed this just now!

**Review by carlocab — APPROVED** [review]

**Pull request closed without merging**